### PR TITLE
Support iterators with incompletely defined __len__ functions

### PR DESCRIPTION
**Pull request by codedecde**
If data_iter has an incomplete definition of __len__, this would throw an error with the check. But as long as the __iter__ function works correctly, the semantics of the data_iterator are unaffected. Making the check more explicit for handling this.

### CODE REVIEW

File: deepspeed/runtime/pipe/engine.py [hunk 1]
```diff
@@ -4,17 +4,14 @@
...
 from deepspeed.utils.timer import ThroughputTimer
 
-from ..engine import DeepSpeedEngine, MEMORY_OPT_ALLREDUCE_SIZE
-from ..utils import PartitionedTensor
 from ..dataloader import RepeatingLoader
-
-from .module import PipelineModule, PipelineError
-from . import p2p
-from . import schedule
+from ..engine import MEMORY_OPT_ALLREDUCE_SIZE, DeepSpeedEngine
```
Comment (tjruwase) [inline deepspeed/runtime/pipe/engine.py:11]: Are these imports needed?

Comment (codedecde) [reply deepspeed/runtime/pipe/engine.py:11]: These were already present. My linter just sorted them.

Comment (loadams) [reply deepspeed/runtime/pipe/engine.py]: Merged changes, ran our formatter

**Comment by loadams** [comment]
@codedecde - is this still a bug that you'd like to see fixed?

**Review by tjruwase — APPROVED** [review]

**Pull request merged**